File: Labs/Lab 8/Solution.py
from typing import List
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        array = [float('inf')] * (amount +1)
        array[0] = 0
        for i in range(1, amount+1):
            for coin in coins:
                if coin <= i:
                    array[i] = min(array[i], 1+array[i-coin])
        
        return array[amount] if array[amount] != float('inf') else -1   
    
solution = Solution()
coins1 = [1, 2, 5]
amount1 = 11
print("Test Case 1:")
print("Input: coins =", coins1, ", amount =", amount1)
print("Output:", solution.coinChange(coins1, amount1))

coins2 = [2]
amount2 = 3
print("\nTest Case 2:")
print("Input: coins =", coins2, ", amount =", amount2)
print("Output:", solution.coinChange(coins2, amount2))

File: Labs/Lab 8/Solution.java
import java.util.Arrays;

public class Solution{
    public int coinChange(int[] coins, int amount) {
        int [] array = new int[amount+1];
        Arrays.fill(array, amount+1);
        array[0] = 0;
        for(int i = 1;i<=amount;i++){
            for(int coin:coins){
                if(coin <= i){
                    array[i]=Math.min(array[i],1+array[i-coin]);
                }
            }
        }
        if(array[amount]<=amount) return array[amount];
        else{
            return -1;
        }
    }
    public static void main(String[] args) {
        Solution solution = new Solution();

        // Test cases
        int[] coins1 = {1, 2, 5};
        int amount1 = 11;
        System.out.println("Test Case 1:");
        System.out.println("Input: coins = " + Arrays.toString(coins1) + ", amount = " + amount1);
        System.out.println("Output: " + solution.coinChange(coins1, amount1));

        int[] coins2 = {2};
        int amount2 = 3;
        System.out.println("\nTest Case 2:");
        System.out.println("Input: coins = " + Arrays.toString(coins2) + ", amount = " + amount2);
        System.out.println("Output: " + solution.coinChange(coins2, amount2));
    }
}